include/RAJA/exec-cilk/forall_cilk.hxx

```diff
@@ -196,128 +196,144 @@ void forall_Icount(cilk_for_exec,
-      loop_body( ii );
-   }
-
-   RAJA_FT_END ;
-}
 
 /*!
  ******************************************************************************
  *
- * \brief  cilk_for iteration over index range with stride with index count.
+ * \brief  cilk_for iteration over raw Box with index count.
  *
  *         NOTE: lambda loop body requires two args (icount, index).
  *
  ******************************************************************************
  */
-template <typename LOOP_BODY>
-RAJA_INLINE
-void forall_Icount(cilk_for_exec,
-                   Index_type begin, Index_type end,
-                   Index_type stride,
-                   Index_type icount,
-                   LOOP_BODY loop_body)
-{
-   Index_type loop_end = (end-begin)/stride;
-   if ( (end-begin) % stride != 0 ) loop_end++;
-
-   RAJA_FT_BEGIN ;
-
-   cilk_for ( Index_type ii = 0 ; ii < loop_end ; ++ii ) {
-      loop_body( ii+icount, begin + ii*stride );
-   }
-
-   RAJA_FT_END ;
-}
-
 
 //
 //////////////////////////////////////////////////////////////////////
 //
-// Function templates that iterate over range-stride segment objects.
+// Function templates that iterate over Box segment objects.
 //
 //////////////////////////////////////////////////////////////////////
 //
 
+#if defined(RAJA_USE_BOXSEGMENT)
 /*!
  ******************************************************************************
  *
- * \brief  cilk_for iteration over range-stride segment object.
+ * \brief  cilk_for iteration over Box segment object.
  *
  ******************************************************************************
  */
+
 template <typename LOOP_BODY>
 RAJA_INLINE
 void forall(cilk_for_exec,
-            const RangeStrideSegment& iseg,
+            const BoxSegment& iseg,
             LOOP_BODY loop_body)
 {
-   Index_type begin  = iseg.getBegin();
-   Index_type end    = iseg.getEnd();
-   Index_type stride = iseg.getStride();
+   Index_type        corner = iseg.getCorner();
+   Index_type        dim    = iseg.getDim();
+   Index_type const *extent = iseg.getExtent();
+   Index_type const *stride = iseg.getStride();
 
    RAJA_FT_BEGIN ;
 
-   cilk_for ( Index_type ii = begin ; ii < end ; ii += stride ) {
-      loop_body( ii );
+   if (dim == 2) {
+      Index_type xExtent = extent[0] ; /* fast stride */
+      Index_type xStride = stride[0] ;
+      Index_type yExtent = extent[1] ; /* slow stride */
+      Index_type yStride = stride[1] ;
+
+      cilk_for ( Index_type y = 0 ; y < yExtent ; ++y ) {
+         Index_type idx = corner + y*yStride ;
+         for (Index_type x = 0 ; x < xExtent ; ++x) {
+            loop_body( idx );
+            idx += xStride ;
+         }
+      }
+   }
+   else if (dim == 1) {
+      Index_type xExtent = extent[0] ; /* fast stride */
+      Index_type xStride = stride[0] ;
+
+      cilk_for (Index_type x = 0 ; x < xExtent ; ++x) {
+         loop_body( corner + x*xStride );
+      }
+   }
+   else /* dim == 3 */ {
```

Comment:
It is an absolute requirement that the 3d be implemented before this branch is merged to develop.  After discussions with you, I realized their are legitimate reasons a person might want to go to higher dimensions.  I still am not sure how to combine a compile time template argument for dimension with a runtime constructor that can collapse that dimension, though your idea of a 'max' dimension might be as good as it gets (although possibly confusing to anyone who dives into the implementation).